### PR TITLE
Add appsec component

### PR DESCRIPTION
**Pull request by lloeki**
**What does this PR do?**

Add an AppSec component, encapsulating the processor independently of the Rack integration.

Another significant change we introduce is that customers no longer need to `require ddtrace/appsec` to enable appsec specific settings.  The autoloading feature of appsec still works as expected. 

When booting ddtrace and when creating the appsec component, if appsec is not enabled, the component will not be instantiated. 

**Motivation**

Enabling AppSec lifecycle management outside of the Rack integration

**Additional Notes**

The value of this component will increase as reconfiguration, notably through remote configuration, becomes a requirement.

**How to test the change?**

Unit tests have been added for the new appsec components and modified existing tests in the core.

### CODE REVIEW

File: sig/datadog/core/configuration/settings.rbs
Comment (lloeki) [inline sig/datadog/core/configuration/settings.rbs:17]: Can this be further typed? I expect this to be `AppSec::Component`.

File: sig/datadog/core/configuration/settings.rbs
Comment (lloeki) [inline sig/datadog/core/configuration/settings.rbs:19]: The `tracer` name seems incorrect, I suspect it is meant to be `component`.

File: lib/datadog/appsec/component.rb
Comment (lloeki) [inline lib/datadog/appsec/component.rb:25]:
```suggestion
      def build_appsec_component(settings, agent_settings, tracer)
```

**Review by lloeki — COMMENTED** [review]
Looks good. Can't approve as I'm the author.

I'm wondering about `Client` though.

File: lib/datadog/appsec/component.rb
Comment (GustavoCaso) [reply lib/datadog/appsec/component.rb:11]: Right, so the component is what gets added to the core components file, following the other libraries
https://github.com/DataDog/dd-trace-rb/blob/b7591eb03d626f5c85b5f5b9e74c530366b86242/lib/datadog/core/configuration/components.rb#L21-L23

Since the `Components` is a module, we can not instantiate it. 

That is why I created a separate class that holds the data for the other app-sec related things: processor, transport, etc... 

I couldn't figure out a better name, so I opted for having inside the component module for know. Happy to find a better name for it. 

Any suggestions? @lloeki

File: lib/datadog/appsec.rb
Comment (GustavoCaso) [reply lib/datadog/appsec.rb:17]: When app sec is not enabled, for what ever reason it was being trigger, that is why I had to make a check to avoid `NoMethodError`

Comment (GustavoCaso) [reply lib/datadog/appsec.rb:17]: My guess is because some tests change the configuration on the fly and doesn't cause the components to rebuild

File: spec/datadog/core/diagnostics/environment_logger_spec.rb
Comment (ivoanjo) [inline spec/datadog/core/diagnostics/environment_logger_spec.rb:41]: I'm curious -- why was this and the other similar change on `buffer_spec.rb` needed?

File: lib/datadog/appsec/component.rb
Comment (lloeki) [reply lib/datadog/appsec/component.rb:11]: Ah, I missed that you changed `Component` to be a module.

Actually the others should become classes, but that's part of a larger scoped refactoring that we're delaying for now, with AppSec leading the design, hence why `AppSec::Component` should be a class here.

Comment (lloeki) [reply lib/datadog/appsec/component.rb:11]: Here's a ref: https://github.com/DataDog/dd-trace-rb/pull/2561#discussion_r1084216245

The idea being that component-specific instances should be owned by each namespace and live inside each component instead of inside `Core::Configuration::Components`.

File: spec/datadog/core/diagnostics/environment_logger_spec.rb
Comment (GustavoCaso) [reply spec/datadog/core/diagnostics/environment_logger_spec.rb:41]: The fact that we are building the appsec component when booting up triggers some calls to the logger when booting up the libddwaf dependency:

https://github.com/DataDog/dd-trace-rb/blob/cebb06754a85397fc88f45ff5f172f5364d6d98a/lib/datadog/appsec/processor.rb#L176

https://github.com/DataDog/dd-trace-rb/blob/cebb06754a85397fc88f45ff5f172f5364d6d98a/lib/datadog/appsec/processor.rb#L50

https://github.com/DataDog/dd-trace-rb/blob/cebb06754a85397fc88f45ff5f172f5364d6d98a/lib/datadog/appsec/processor.rb#L182

The tests stubs only the `:info` method so the test was failing with unexpected method call

Comment (GustavoCaso) [reply spec/datadog/core/diagnostics/environment_logger_spec.rb:41]: We could have a more ergonomic test helper for mocking `Datadog.logger`.

That could be incorporated into a separate PR.

Comment (ivoanjo) [reply spec/datadog/core/diagnostics/environment_logger_spec.rb:41]: Ack :+1: 

To be honest, a lot of these tests boot up half of the library, which I think they shouldn't (rather than just fixing the logger), but I do agree some of those changes are probably better done as a separate :)

File: spec/datadog/appsec/component_spec.rb
Comment (ivoanjo) [inline spec/datadog/appsec/component_spec.rb:32]: This seems to be a bit of a code smell of this class being a bit hard to test. Consider perhaps making the `Processor` a dependency that can be injected?

```ruby
def initialize(settings:, processor: Processor.new)
```

and then you can supply your own processor here. Furthermore, consider using a `instance_double` for the `Processor`, instead of just a naked `double`, since that means that means that RSpec will check that the class you mentioned actually exists and actually has the `finalize` method and that it doesn't receive arguments (aka that it exactly matches the double).

File: spec/datadog/appsec/component_spec.rb
Comment (ivoanjo) [inline spec/datadog/appsec/component_spec.rb:45]: See my note above on how I suggest changing the class to make it more testable and avoid needing to do this :)

File: spec/datadog/core/configuration/components_spec.rb
Comment (ivoanjo) [inline spec/datadog/core/configuration/components_spec.rb:1392]: If you do decide to keep the replacement support (don't :P), this test description is incorrect ;)

But better yet -- delete all this! :D

**Review by ivoanjo — COMMENTED** [review]
Left a few notes, nothing big!

File: lib/datadog/appsec.rb
Comment (GustavoCaso) [reply lib/datadog/appsec.rb:18]: Thanks for the detailed explanation.

Yes, I had `components.appsec&.processor` before, but CO for old rubies 💣  so I had to change it. 

I'll hoist the component in a local variable

File: lib/datadog/appsec.rb
Comment (GustavoCaso) [reply lib/datadog/appsec.rb:14]: I will move it to the private block. Also, to be fair, nothing within the appsec module is marked as public

File: lib/datadog/appsec.rb
Comment (GustavoCaso) [reply lib/datadog/appsec.rb:35]: I guess the part that is not necessary anymore is the `require 'ddtrace'` within the autoload file. Since we could assume `ddtrace` has been loaded before 😄

File: lib/datadog/appsec.rb
Comment (ivoanjo) [reply lib/datadog/appsec.rb:18]: Feel free to update the "OLD RUBY PAIN TRACKER" spreadsheet linked on the internal Ruby channel with this instance, as it's an example of something where you lost time going back and forth (and me commenting on it) because we're still supporting those Rubies.

File: lib/datadog/appsec.rb
Comment (GustavoCaso) [reply lib/datadog/appsec.rb:23]: The `writer` is never used within the appsec module. We might as well remove it altogether. 

@lloeki any thoughts?

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (GustavoCaso) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:26]: I can look into doing that. If I see the changes are bigger than expected I will move the work to a separate PR

File: lib/datadog/core/configuration/components.rb
Comment (GustavoCaso) [reply lib/datadog/core/configuration/components.rb:117]: Great point 👍

File: lib/datadog/appsec.rb
Comment (ivoanjo) [reply lib/datadog/appsec.rb:14]: > Also, to be fair, nothing within the appsec module is marked as public

Right... You are right that public methods are "supposed to" be marked with `@public_api`, although I would argue that we should be extra cautious about what we accidentally expose in the top-level product modules, because of expectations set in other components -- e.g. everything "public" at the Ruby level in `Datadog::Tracing`, `Datadog::Profiling` is supposed to be a `@public_api`, so it may lead customers to assume the same was true for `Datadog::AppSec`.

File: sig/datadog/core/configuration/agent_settings_resolver.rbs
Comment (GustavoCaso) [reply sig/datadog/core/configuration/agent_settings_resolver.rbs:5]: I will leave that exercise for me on a future PR :)

File: spec/datadog/appsec/component_spec.rb
Comment (GustavoCaso) [reply spec/datadog/appsec/component_spec.rb:10]: I agree with that. I initialled started using `AppSec::Configuration::Settings`. The problem is there is no way to override the values of the settings unless using ENV variables, which would mean I would have to mock `ENV`, which is fine, but I thought just not doing that and going the easy path. 

I will add a note to myself to improve `AppSec::Configuration::Settings` to allow to accept some kind of overrides so it would improve testing in the future. 

In the meantime, I will change the test to use `AppSec::Configuration::Settings`

Comment (ivoanjo) [reply spec/datadog/appsec/component_spec.rb:10]: I guess `Datadog::AppSec::Configuration::Settings.new.merge(Datadog::AppSec::Configuration::DSL.new.tap { |it| it.enabled = true })` works but it does have a bit of a code smell :)

File: lib/datadog/appsec.rb
Comment (ivoanjo) [inline lib/datadog/appsec.rb:20]: Following from https://github.com/DataDog/dd-trace-rb/pull/2632#discussion_r1114143568 -- I guess do quickly doublecheck if `#processor` should be considered public or not.

File: spec/datadog/appsec/component_spec.rb
Comment (GustavoCaso) [reply spec/datadog/appsec/component_spec.rb]: I simplify it by doing the following: 

```ruby
it 'returns a Datadog::AppSec::Component instance' do
  Datadog.configuration.appsec.enabled = true
  component = described_class.build_appsec_component(Datadog.configuration)
  expect(component).to be_a(described_class)
end
```

File: spec/datadog/appsec/component_spec.rb
Comment (ivoanjo) [inline spec/datadog/appsec/component_spec.rb:18]: I know you adopted this after my previous suggestion, but be careful about solving it in this way because this is state that gets carried over to other tests.

Tests that change the global configuration should make sure to call 
`Datadog.configuration.reset!` at the end (you'll see this in other places in the codebase where we do this.

Although it's kinda uglier, I do still favor my suggestion from https://github.com/DataDog/dd-trace-rb/pull/2632#discussion_r1114180817 as at least that doesn't leave any lingering state, but I'll leave it up to you to decide which approach to better use (just make sure no state gets left! :D).

File: spec/datadog/appsec/component_spec.rb
Comment (ivoanjo) [inline spec/datadog/appsec/component_spec.rb:32]: Minor: Consider replacing this with an `instance_double` -- that would avoid the need to stub out stuff on the real object, while still giving us all of the same safety checks.

File: spec/datadog/appsec/component_spec.rb
Comment (ivoanjo) [inline spec/datadog/appsec/component_spec.rb:38]: Minor:

```suggestion
      it 'does not finalize the processor' do
```

File: spec/datadog/appsec/component_spec.rb
Comment (ivoanjo) [inline spec/datadog/appsec/component_spec.rb:41]: Minor: Personally, I don't think this expectation is needed. The way I see it, it's ok in this case to use our knowledge of the implementation to validate that this doesn't happen.

You'd need to be doing something really really weird for this expectation to ever fail, and if we go into the realm of "really really weird", then we could be asserting on a lot of other things that we aren't -- maybe it's allocating 10000 empty objects? maybe the component is mining bitcoin? maybe it's ordering me a nice pizza?

**Review by ivoanjo — APPROVED** [review]
:+1: Left a few final suggestions, and overall it LGTM

File: lib/datadog/appsec.rb
Comment (GustavoCaso) [reply lib/datadog/appsec.rb:20]: We use it here 
https://github.com/DataDog/dd-trace-rb/blob/99052d9ec16b3a9505a25807bdd201b5f3b0dd58/lib/datadog/appsec/contrib/rack/request_middleware.rb#L28

File: spec/datadog/appsec/component_spec.rb
Comment (GustavoCaso) [reply spec/datadog/appsec/component_spec.rb:18]: Sounds good

File: lib/datadog/appsec.rb
Comment (ivoanjo) [reply lib/datadog/appsec.rb:20]: We actually have avoided still having these as "Ruby public" in a few places in the codebase -- you'll find a few `.send(...)` if you look for them, which is our attempt at prioritizing the customer communication part ("you shouldn't rely on this") vs the library ergonomics part ("send(...) is really ugly").

But... I don't have strong feelings on this, we're definitely not consistent in our approach here.

File: spec/datadog/appsec/component_spec.rb
Comment (GustavoCaso) [reply spec/datadog/appsec/component_spec.rb:18]: We should probably add some tracking for this kind of pattern. I have seen in other tests thats why I used it.

File: sig/datadog/core/configuration/agent_settings_resolver.rbs
Comment (lloeki) [reply sig/datadog/core/configuration/agent_settings_resolver.rbs:5]: Note: as we add more things to `sig/datadog/{core,tracing,...}`, we may want to add some `check ...` in `Steepfile` so that their corresponding source also gets checked.

File: lib/datadog/appsec.rb
Comment (lloeki) [reply lib/datadog/appsec.rb:31]: @GustavoCaso Remove it! It's unused and won't ever be.

Context: it was used when AppSec events were sent over their own API (and not as tags on the trace), which is a design that was discovered as being improper for multiple reasons.

As a general rule, if you stumble upon unused files/code, feel free to open a dedicated PR to cull those out.

File: lib/datadog/appsec.rb
Comment (lloeki) [reply lib/datadog/appsec.rb:43]: @GustavoCaso I presume you're correct.

That said while I don't think `require 'datadog/appsec/autoload'` was ever exposed to end users we should be extra careful and check the docs if that's the case anywhere.

Comment (lloeki) [reply lib/datadog/appsec.rb:43]: Also now that we don't gate out requiring `datadog/appsec`, I think `lib/datadog/appsec/autoload.rb` should ultimately go away, and `DD_APPSEC_ENABLED=true` should be handled via the regular configuration only.

There would be a bunch of questions remaining regarding auto-instrumenting AppSec as it is done currently when `DD_APPSEC_ENABLED=true` but not when `Datadog.configure { |c| c.appsec.enabled = true }`.

File: spec/datadog/appsec/component_spec.rb
Comment (lloeki) [inline spec/datadog/appsec/component_spec.rb:12]: Minor: this `it` is confusing, as it is not the same `it` as RSpec's DSL, nor the `subject` of the test case. I feel this would have been more readable:

```
          Datadog::AppSec::Configuration::DSL.new.tap do |appsec|
            appsec.enabled = true
```

**Pull request merged**
